pkg/wasm/cache.go

```diff
@@ -0,0 +1,194 @@
+	"net/url"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"istio.io/pkg/log"
+)
+
+var wasmLog = log.RegisterScope("wasm", "", 0)
+
+const (
+	// DefaultWasmModulePurgeInteval is the default interval for periodic stale Wasm module clean up.
+	DefaultWasmModulePurgeInteval = 10 * time.Minute
+
+	// DefaultWasmModuleExpiry is the default duration for least recently touched Wasm module to become stale.
+	DefaultWasmModuleExpiry = 24 * time.Hour
+)
+
+// Cache models a Wasm module cache.
+// TODO(bianpengyuan): add metrics to the wasm package.
+type Cache interface {
+	Get(url, checksum string, timeout time.Duration) (string, error)
+}
+
+// LocalFileCache for downloaded Wasm modules. Currently it stores the Wasm module as local file.
+type LocalFileCache struct {
+	// Map from Wasm module checksum to cache entry.
+	modules map[cacheKey]cacheEntry
+
+	// http fetcher fetches Wasm module with HTTP get.
+	httpFetcher *HTTPFetcher
+
+	// directory path used to store Wasm module.
+	dir string
+
+	// mux is needed because stale Wasm module files will be purged periodically.
+	mux sync.Mutex
+
+	// Duration for stale Wasm module purging.
+	purgeInterval    time.Duration
+	wasmModuleExpiry time.Duration
+
+	stopChan chan struct{}
+}
+
+var _ Cache = &LocalFileCache{}
+
+type cacheKey struct {
+	downloadURL string
+	checksum    string
+}
+
+// cacheEntry contains information about a Wasm module cache entry.
+type cacheEntry struct {
+	// File path to the downloaded wasm modules.
+	modulePath string
+
+	// Last time that this local Wasm module is referenced.
+	last time.Time
+}
+
+// NewLocalFileCache create a new Wasm module cache which downloads and stores Wasm module files locally.
+func NewLocalFileCache(dir string, purgeInterval, moduleExpiry time.Duration) *LocalFileCache {
+	cache := &LocalFileCache{
+		httpFetcher:      NewHTTPFetcher(),
+		modules:          make(map[cacheKey]cacheEntry),
+		dir:              dir,
+		purgeInterval:    purgeInterval,
+		wasmModuleExpiry: moduleExpiry,
+		stopChan:         make(chan struct{}),
+	}
+	go func() {
+		cache.purge()
+	}()
+	return cache
+}
+
+// Get returns path the local Wasm module file.
+func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration) (string, error) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	url, err := url.Parse(downloadURL)
+	if err != nil {
+		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
+	}
+	// Construct Wasm cache key with downloading URL and provided checksum of the module.
+	key := cacheKey{
+		downloadURL: downloadURL,
+		checksum:    checksum,
+	}
+
+	switch url.Scheme {
+	case "http", "https":
+		// First check if the cache entry is already downloaded.
+		if ce, ok := c.modules[key]; ok {
+			// Update last touched time.
+			ce.last = time.Now()
+			return ce.modulePath, nil
+		}
+
+		// If the module is not available locally, download the Wasm module with http fetcher.
+		b, err := c.httpFetcher.Fetch(downloadURL, timeout)
```

Comment:
nit: might be overthinking, but this fetching is now single threaded because of mutex. Would be better to allow parallel fetches (but can be in a followup).

Reply:
Yeah good point, thanks! I was thinking about wrapping map access in getter and setter while writing this, but forgot to add it. I updated the change to add that.